Implement secp256k1-verify, secp25k1-recover, principal-of

psq opened the pull request:
This PR will include the implementation for the Clarity functions to verify secp256k1 signatures.  See #1134 for details.

Starting with upgrading to latest version of the library.

psq commented:
the initial proposal from @lgalabru calls for the `recover` function to have the this signature:
```
(secp256k1-recover (buff 32) (buff 65)) -> (response principal uint)
```

However, if you'd like to turn around and use the recovered key with `secp256k1-verify`, you couldn't as you can not get the public key from the principal.  So instead, I'm considering implementing:
```
(secp256k1-recover (buff 32) (buff 65)) -> (response (buff 33) uint)
```
If you need the principal, you can get it using the suggested `(principal-of (buff 33)) -> (response principal uint)`, which makes recovering the key as a `principal` a 2 step process, but you also get the public key as a `(buff 33)`.

Any thoughts? objections?

psq commented:
there was still in an older version of the secp256k1 crate in the stacks-node Cargo.toml.  Did not seem to be used, as removing it still builds...

kantai commented on src/vm/functions/mod.rs:
```diff
@@ -397,3 +408,113 @@ fn special_contract_of(args: &[SymbolicExpression], env: &mut Environment, conte
+    Ok(Value::Principal(principal))
+}
+
+fn special_secp256k1_recover(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
```
I think we should put these functions in a module rather than adding them here (perhaps `arithmetic`?)

psq replied:
I followed the lead of how/where `sha...` and `keccak...` were done.  Not sure, these would belong in `arithmetic` but maybe all these could be moved into a new module named `crypto` maybe?

kantai replied:
Yeah, a new module `crypto` makes sense

psq replied:
see 3907ed5

kantai commented on src/vm/functions/mod.rs:
```diff
@@ -397,3 +408,113 @@ fn special_contract_of(args: &[SymbolicExpression], env: &mut Environment, conte
+                return Err(CheckErrors::TypeValueError(BUFF_65, param1).into())
+            }
+            if data[64] > 3 {
+                return Err(CheckErrors::InvalidSecp65k1Signature.into())
```
I think this should also be a handle-able error code in Clarity (i.e., return `Value::err_uint(2)`), and for the data.len() check, it should only be a TypeValueError if `data.len() > 65` (that's all that the type system will enforce), for `data.len() < 65`, we should probably have `Value::err_uint(2)` as well.

psq replied:
sure, I'll make the change.  The `verify` function should behave the way as well, then.

psq replied:
address in 3907ed5, except I changed verify to return `false` instead of throwing an error

lgalabru commented on src/vm/analysis/type_checker/natives/mod.rs:
```diff
@@ -288,6 +288,27 @@ fn check_contract_of(checker: &mut TypeChecker, args: &[SymbolicExpression], con
+fn check_principal_of(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(1, args)?;
+    checker.type_check_expects(&args[0], context, &BUFF_33)?;
+    Ok(TypeSignature::PrincipalType)
```
Building on our last meeting, this function should be returning a PrincipalType wrapped in an Optional / Response instead.

psq replied:
changed in 3907ed5

lgalabru commented on src/vm/analysis/type_checker/natives/mod.rs:
```diff
@@ -288,6 +288,27 @@ fn check_contract_of(checker: &mut TypeChecker, args: &[SymbolicExpression], con
+    checker.type_check_expects(&args[0], context, &BUFF_32)?;
+    checker.type_check_expects(&args[1], context, &BUFF_65)?;
+    checker.type_check_expects(&args[2], context, &BUFF_33)?;
+    Ok(TypeSignature::BoolType)
```
Along the same vein, should we wrapped this `BoolType` in an Optional / Response, so developers have a way to handle a signature failing at verifying, and a bad construct.

psq replied:
hmm, I thought I had commented on this already.  I can't figure out a use case where a contract could recover from a signature not verifying for whatever reason.  The only thing that matters is whether the function returns true or not.  Based on some other feedback, it will return false more than before, allowing the contract to maybe try an other signature and ignore that one...

kantai replied:
Returning true/false works for me -- the only reason we may want to use a response type is if we want err codes for the signature failures.

lgalabru commented on src/vm/functions/mod.rs:
```diff
@@ -57,10 +62,13 @@ define_named_enum!(NativeFunctions {
     ContractCall("contract-call?"),
     AsContract("as-contract"),
     ContractOf("contract-of"),
+    PrincipalOf("principal-of"),
```
Related to previous feedbacks, the functions returning Optional / Response should end with a `?` (`principal-of?`, etc)

psq replied:
addressed with 3907ed5

psq commented:
I thought the bitcoind test failures were a fluke, but that was actually not the case and glad I fixed it.

@kantai @lgalabru I think I've addressed all the feedback, anything else needed?

lgalabru commented on src/vm/docs/mod.rs:
```diff
@@ -619,6 +619,36 @@ is supplied the hash is computed over the little-endian representation of the in
+};
+
+const SECP256K1VERIFY_API: SpecialAPI = SpecialAPI {
+    input_type: "(buff 32), (buff 65), (buff 33)",
```
I missed this the first time. In this case, we're verifying, meaning we should be dealing with a "standard" signature `buff 64` instead of a `buff 65`, no?

psq replied:
this is intentional so the same signature (which includes the recovery id as the last byte) can be used for both `secp256k1-verify` and `secp256k1-recover`.  If the hash was signed with `sign` and not `sign_recoverable`,  `0x00` can be added.

We could make it so that both 64 bytes and 65 bytes could be passed to `secp256k1-verify` but the 64 bytes version would fail if you pass it to `secp356k1-recover`.

I think just having to use 65 bytes keeps things simpler.

lgalabru replied:
In my experience, either you sign for verify, or you sign(_recoverable) for recover, but I'm having a hard time envisioning a use case in a smart contract, where you would use a verify and recover in the same sequence, and with the same signature.
In my mind, it's easier and simpler for a developer to just throw 1) the output of `sign` to `verify`, and 2) the output of `sign_recoverable` to `recover`, vs adding one useless extra byte in case 1.

psq replied:
I changed the function so it can take either a 64 or 65 bytes signatures, which should worth whether the signature includes the recoveryId or not (see 6db1099a446a963d45655f2ab59028fb1b32e19f)

lgalabru reviewed (approved):
Looks good to me, thank you @psq, super excited by the new feature!
Referencing https://github.com/clarity-lang/reference/issues/4 for completion

kantai reviewed (approved):
LGTM! Merging.

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
